esp-hal/src/system.rs

```diff
@@ -122,234 +134,225 @@ pub enum Peripheral {
     Systimer,
 }
 
+static PERIPHERAL_REF_COUNT: [AtomicUsize; Peripheral::COUNT] =
+    [const { AtomicUsize::new(0) }; Peripheral::COUNT];
+
+/// Disable all peripherals.
+///
+/// Peripherals listed in [KEEP_ENABLED] are NOT disabled.
+pub(crate) fn disable_peripherals() {
+    for p in Peripheral::iter() {
+        if KEEP_ENABLED.contains(&p) {
+            continue;
+        }
+        PeripheralClockControl::enable_forced(p, false, true);
+    }
+}
+
+pub(crate) struct PeripheralGuard {
+    peripheral: Peripheral,
+}
+
+impl PeripheralGuard {
+    pub(crate) fn new(p: Peripheral) -> Self {
+        if !KEEP_ENABLED.contains(&p) && PeripheralClockControl::enable(p, true) {
+            PeripheralClockControl::reset(p);
+        }
+
+        Self { peripheral: p }
+    }
+}
+
+impl Drop for PeripheralGuard {
+    fn drop(&mut self) {
+        if !KEEP_ENABLED.contains(&self.peripheral) {
+            PeripheralClockControl::enable(self.peripheral, false);
+        }
+    }
+}
+
 /// Controls the enablement of peripheral clocks.
 pub(crate) struct PeripheralClockControl;
 
 #[cfg(not(any(esp32c6, esp32h2)))]
 impl PeripheralClockControl {
-    /// Enables and resets the given peripheral
-    pub(crate) fn enable(peripheral: Peripheral) {
-        let system = unsafe { &*SYSTEM::PTR };
 
-        #[cfg(esp32)]
-        let (perip_clk_en0, perip_rst_en0, peri_clk_en, peri_rst_en) = {
-            (
-                &system.perip_clk_en(),
-                &system.perip_rst_en(),
-                &system.peri_clk_en(),
-                &system.peri_rst_en(),
-            )
-        };
-        #[cfg(not(esp32))]
-        let (perip_clk_en0, perip_rst_en0) = { (&system.perip_clk_en0(), &system.perip_rst_en0()) };
+    fn enable_internal(peripheral: Peripheral, enable: bool) {
+        debug!("Enable {:?} {}", peripheral, enable);
 
-        #[cfg(any(esp32c2, esp32c3, esp32s2, esp32s3))]
-        let (perip_clk_en1, perip_rst_en1) = { (&system.perip_clk_en1(), &system.perip_rst_en1()) };
+        critical_section::with(|_cs| {
```

Comment:
it's currently only called via `enable_forced` so we are already in a critical-section - not sure if it's ok to remove it here and have a comment (and maybe mark the function unsafe)

Reply:
You can also change `force` into `Option<&CriticalSection>`

Reply:
Not sure I understand that - I guess ``enable_forced` could add a paremeter `&CriticalSection` (not optional) but also the C6/H2 implementation should take it then .... I guess I will do that

Reply:
Oh hm this is a bit hard to see in the diff view, so I missed where this comment was. If this function is only called in a critical section, just take a critical section (and don't bother with the forced part of my comment).